Mathlib/RingTheory/LocalRing/Module.lean

```diff
@@ -0,0 +1,270 @@
+`M` finite and `N` finite free, then `k ⊗ M → k ⊗ N` is injective implies `P` is free.
+-/
+theorem free_of_lTensor_residueField_injective
+    [Module.Finite R M] [Module.Finite R N] [Module.Free R N]
+    (hf : Function.Injective (f.lTensor k)) :
+    Module.Free R P := by
+  have := Module.finitePresentation_of_free_of_surjective g hg
+    (by rw [h.linearMap_ker_eq, LinearMap.range_eq_map]; exact (Module.Finite.out).map f)
+  apply free_of_maximalIdeal_rTensor_injective
+  rw [← LinearMap.lTensor_inj_iff_rTensor_inj]
+  apply lTensor_injective_of_exact_of_exact_of_rTensor_injective
+    h hg (LinearMap.exact_subtype_mkQ 𝔪) (Submodule.mkQ_surjective _)
+    ((LinearMap.lTensor_inj_iff_rTensor_inj _ _).mp hf)
+    (Module.Flat.lTensor_preserves_injective_linearMap _ Subtype.val_injective)
+
+end Module
+
+/--
+Given a linear map `l : M → N` over a local ring `(R, 𝔪, k)`
+with `M` finite and `N` is finite free,
+`l` is a split injection iff `k ⊗ l` is a (split) injection.
+-/
+theorem LocalRing.split_injective_iff_lTensor_residueField_injective
+    [Module.Finite R M] [Module.Finite R N] [Module.Free R N] (l : M →ₗ[R] N) :
+    (∃ l', l' ∘ₗ l = LinearMap.id) ↔ Function.Injective (l.lTensor (ResidueField R)) := by
+  constructor
+  · intro ⟨l', hl⟩
+    have : l'.lTensor (ResidueField R) ∘ₗ l.lTensor (ResidueField R) = .id := by
+      rw [← LinearMap.lTensor_comp, hl, LinearMap.lTensor_id]
+    exact Function.HasLeftInverse.injective ⟨_, LinearMap.congr_fun this⟩
+  · intro h
+    -- By the lemma above, `k ⊗ l` injective => `N ⧸ l(M)` free.
+    have := Module.free_of_lTensor_residueField_injective l (LinearMap.range l).mkQ
+      (Submodule.mkQ_surjective _) l.exact_map_mkQ_range h
+    -- Hence `0 → l(M) → N → N ⧸ l(M) → 0` splits and `l(M)` is projective.
+    have : Module.Projective R (LinearMap.range l) := by
+      have := (Exact.split_tfae (LinearMap.exact_subtype_mkQ (LinearMap.range l))
+        Subtype.val_injective (Submodule.mkQ_surjective _)).out 0 1
+      obtain ⟨l', hl'⟩ := this.mp
+          (Module.projective_lifting_property _ _ (Submodule.mkQ_surjective _))
+      exact Module.Projective.of_split _ _ hl'
+    -- Then `0 → ker l → M → l(M) → 0` splits.
+    obtain ⟨l', hl'⟩ : ∃ l', l' ∘ₗ (LinearMap.ker l).subtype = LinearMap.id := by
+      have : Function.Exact (LinearMap.ker l).subtype
+          (l.codRestrict (LinearMap.range l) (LinearMap.mem_range_self l)) := by
+        rw [LinearMap.exact_iff, LinearMap.ker_rangeRestrict, Submodule.range_subtype]
+      have := (Exact.split_tfae this
+        Subtype.val_injective (fun ⟨x, y, e⟩ ↦ ⟨y, Subtype.ext e⟩)).out 0 1
+      exact this.mp (Module.projective_lifting_property _ _ (fun ⟨x, y, e⟩ ↦ ⟨y, Subtype.ext e⟩))
+    have : Module.Finite R (LinearMap.ker l) := by
+      refine Module.Finite.of_surjective l' ?_
+      exact Function.HasRightInverse.surjective ⟨_, DFunLike.congr_fun hl'⟩
+    -- And `0 → k ⊗ ker l → k ⊗ M → k ⊗ l(M) → 0` also splits.
```

Comment:
Can you explain this point a bit more? The next declaration `H` is just an injectivity, while here (and later on) the splitting plays a role. Perhaps you could introduce `l'` in the previous comment as the map `l(M) -> M` and here say that this injective map stays injective after `k ⊗ -`.

Reply:
The logic here is `0 → ker l → M → l(M) → 0` splits => `k ⊗ ker l → k ⊗ M → k ⊗ l(M) → 0` splits => `k ⊗ ker l → k ⊗ M` injective. And we only need injectivity later on.

Reply:
Oh, I see. Then I would rather say in this comment that since the first sequence splits, tensoring with `k` preserves injectivity of the first arrow in the second sequence. (which is `H`).